titus-common/src/main/java/com/netflix/titus/common/framework/simplereconciler/internal/EventDistributor.java

```diff
@@ -0,0 +1,270 @@
+                snapshot = snapshotSupplier.get();
+            }
+            eventQueue.drainTo(events);
+            eventQueueDepth.accumulateAndGet(events.size(), (current, delta) -> current - delta);
+
+            // We emit events to already connected sinks first. For new sinks, we build snapshot and merge it with the
+            // pending event queue in addNewSinks method.
+            processEvents(events);
+            removeUnsubscribedSinks();
+            if (snapshot != null) {
+                addNewSinks(snapshot, events);
+            }
+            metricEmittedEvents.increment(events.size());
+
+            metricLoopExecutionTime.record(start.elapsed(TimeUnit.MILLISECONDS), TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void addNewSinks(List<SimpleReconcilerEvent<DATA>> snapshot, List<SimpleReconcilerEvent<DATA>> events) {
+        ArrayList<EmitterHolder> newEmitters = new ArrayList<>();
+        sinkQueue.drainTo(newEmitters);
+        if (!newEmitters.isEmpty()) {
+            List<SimpleReconcilerEvent<DATA>> merged = mergeSnapshotAndPendingEvents(snapshot, events);
+            newEmitters.forEach(holder -> {
+                activeSinks.put(holder.getId(), holder);
+                holder.onNext(merged);
+            });
+        }
+    }
+
+    private void removeUnsubscribedSinks() {
+        Set<String> cancelled = new HashSet<>();
+        activeSinks.forEach((id, holder) -> {
+            if (holder.isCancelled()) {
+                cancelled.add(id);
+            }
+        });
+        activeSinks.keySet().removeAll(cancelled);
+    }
+
+    private List<SimpleReconcilerEvent<DATA>> mergeSnapshotAndPendingEvents(List<SimpleReconcilerEvent<DATA>> snapshot,
+                                                                            List<SimpleReconcilerEvent<DATA>> events) {
+        if (events.isEmpty()) {
+            return snapshot;
+        }
+        Map<String, SimpleReconcilerEvent<DATA>> eventsById = new HashMap<>();
+        events.forEach(event -> eventsById.put(event.getId(), event));
+        List<SimpleReconcilerEvent<DATA>> mergeResult = new ArrayList<>();
+        for (SimpleReconcilerEvent<DATA> event : snapshot) {
+            SimpleReconcilerEvent<DATA> fromQueue = eventsById.get(event.getId());
+            if (fromQueue == null || fromQueue.getTransactionId() < event.getTransactionId()) {
+                mergeResult.add(event);
+            } else if (fromQueue.getKind() != SimpleReconcilerEvent.Kind.Removed) {
+                mergeResult.add(new SimpleReconcilerEvent<>(
+                        SimpleReconcilerEvent.Kind.Added,
+                        fromQueue.getId(),
+                        fromQueue.getData(),
+                        fromQueue.getTransactionId()
+                ));
+            }
+        }
+        return mergeResult;
+    }
+
+    private void processEvents(List<SimpleReconcilerEvent<DATA>> events) {
+        events.forEach(event -> {
```

Comment:
I'm a little confused about why we iterate over each individual event here. If `events` has 5 `SimpleReconcilerEvent`s won't we add this same `events` to the `holder.onNext(events)` 5 times?

Reply:
Copy-paste. This class is based on the old reconciler code. Thanks for catching this